File: 515-find-largest-value-in-each-tree-row/find-largest-value-in-each-tree-row.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> largestValues(TreeNode* root) {
        // 24.10.23 POTD
        // APPROACH: BFS Using Queue

        if(!root)   return {};
        
        vector<int> ans;
        queue<TreeNode*> q;

        q.push(root);
        while(q.size()){
            int breadth = q.size();
            int mx = INT_MIN;

            while(breadth--){
                auto node = q.front();
                q.pop();

                mx = max(mx, node->val);

                if(node->left)  q.push(node->left);
                if(node->right) q.push(node->right);
            }

            ans.push_back(mx);
        }
        return ans;
    }
};

File: 515-find-largest-value-in-each-tree-row/find-largest-value-in-each-tree-row.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def largestValues(self, root: Optional[TreeNode]) -> List[int]:
        # 25.12.24 POTD
        if not root:
            return []

        ans, q = [], []
        q.append(root)

        while q:
            sz = len(q)
            mx = q[0].val

            while sz:
                sz -= 1
                nd = q.pop(0)
                mx = max(mx, nd.val)

                if nd.left:
                    q.append(nd.left)
                if nd.right:
                    q.append(nd.right)
            
            ans.append(mx)

        return ans